File: sdks/swift/Mentat/Mentat/Mentat.swift
/* Copyright 2018 Mozilla
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License. */

import Foundation

import MentatStore

public typealias Entid = Int64

/**
 Protocol to be implemented by any object that wishes to register for transaction observation
 */
public protocol Observing {
    func transactionDidOccur(key: String, reports: [TxChange])
}

/**
 Protocol to be implemented by any object that provides an interface to Mentat's transaction observers.
 */
public protocol Observable {
    func register(key: String, observer: Observing, attributes: [String])
    func unregister(key: String)
}

public enum CacheDirection {
    case forward;
    case reverse;
    case both;
}

/**
 The primary class for accessing Mentat's API.
 This class provides all of the basic API that can be found in Mentat's Store struct.
 The raw pointer it holds is a pointer to a Store.
*/
open class Mentat: RustObject {
    fileprivate static var observers = [String: Observing]()

    /**
     Create a new Mentat with the provided pointer to a Mentat Store
     - Parameter raw: A pointer to a Mentat Store.
    */
    public required override init(raw: OpaquePointer) {
        super.init(raw: raw)
    }

    /**
     Open a connection to a Store in a given location.
     If the store does not already exist, one will be created.

     - Parameter storeURI: The URI as a String of the store to open.
        If no store URI is provided, an in-memory store will be opened.
    */
    public class func open(storeURI: String = "") throws -> Mentat {
        return Mentat(raw: try RustError.unwrap({err in store_open(storeURI, err) }))
    }

    /**
     Add an attribute to the cache. The {@link CacheDirection} determines how that attribute can be
     looked up.

     - Parameter attribute: The attribute to cache
     - Parameter direction: The direction the attribute should be keyed.
        `forward` caches values for an attribute keyed by entity
        (i.e. find values and entities that have this attribute, or find values of attribute for an entity)
        `reverse` caches entities for an attribute keyed by value.
        (i.e. find entities that have a particular value for an attribute).
        `both` adds an attribute such that it is cached in both directions.

     - Throws: `ResultError.error` if an error occured while trying to cache the attribute.
     */
    open func cache(attribute: String, direction: CacheDirection) throws {
        try RustError.withErrorCheck({err in
            switch direction {
            case .forward:
                store_cache_attribute_forward(self.raw, attribute, err)
            case .reverse:
                store_cache_attribute_reverse(self.raw, attribute, err)
            case .both:
                store_cache_attribute_bi_directional(self.raw, attribute, err)
            }
        });
    }

    /**
    Simple transact of an EDN string.
     - Parameter transaction: The string, as EDN, to be transacted

     - Throws: `ResultError.error` if the an error occured during the transaction, or the TxReport is nil.

     - Returns: The `TxReport` of the completed transaction
    */
    open func transact(transaction: String) throws -> TxReport {
        return TxReport(raw: try RustError.unwrap({err in store_transact(self.raw, transaction, err) }))
    }

    /**
     Start a new transaction.

     - Throws: `ResultError.error` if the creation of the transaction fails.
     - Throws: `ResultError.empty` if no `InProgress` is created.

     - Returns: The `InProgress` used to manage the transaction
     */
    open func beginTransaction() throws -> InProgress {
        return InProgress(raw: try RustError.unwrap({err in store_begin_transaction(self.raw, err) }));
    }

    /**
     Creates a new transaction (`InProgress`) and returns an `InProgressBuilder` for that transaction.

     - Throws: `ResultError.error` if the creation of the transaction fails.
     - Throws: `ResultError.empty` if no `InProgressBuilder` is created.

     - Returns: an `InProgressBuilder` for this `InProgress`
     */
    open func entityBuilder() throws -> InProgressBuilder {
        return InProgressBuilder(raw: try RustError.unwrap({err in store_in_progress_builder(self.raw, err) }))
    }

    /**
     Creates a new transaction (`InProgress`) and returns an `EntityBuilder` for the entity with `entid`
    for that transaction.

     - Parameter entid: The `Entid` for this entity.

     - Throws: `ResultError.error` if the creation of the transaction fails.
     - Throws: `ResultError.empty` if no `EntityBuilder` is created.

     - Returns: an `EntityBuilder` for this `InProgress`
     */
    open func entityBuilder(forEntid entid: Entid) throws -> EntityBuilder {
        return EntityBuilder(raw: try RustError.unwrap({err in
            store_entity_builder_from_entid(self.raw, entid, err) }))
    }

    /**
     Creates a new transaction (`InProgress`) and returns an `EntityBuilder` for a new entity with `tempId`
    for that transaction.

     - Parameter tempId: The temporary identifier for this entity.

     - Throws: `ResultError.error` if the creation of the transaction fails.
     - Throws: `ResultError.empty` if no `EntityBuilder` is created.

     - Returns: an `EntityBuilder` for this `InProgress`
     */
    open func entityBuilder(forTempId tempId: String) throws -> EntityBuilder {
        return EntityBuilder(raw: try RustError.unwrap({err in
            store_entity_builder_from_temp_id(self.raw, tempId, err)
        }))
    }

    /**
     Get the the `Entid` of the attribute.

     - Parameter attribute: The string represeting the attribute whose `Entid` we are after.
     The string is represented as `:namespace/name`.

     - Returns: The `Entid` associated with the attribute.
     */
    open func entidForAttribute(attribute: String) -> Entid {
        return Entid(store_entid_for_attribute(self.raw, attribute))
    }

    /**
     Start a query.
     - Parameter query: The string represeting the the query to be executed.

     - Returns: The `Query` representing the query that can be executed.
     */
    open func query(query: String) -> Query {
        return Query(raw: store_query(self.raw, query))
    }

    /**
     Retrieve a single value of an attribute for an Entity
     - Parameter attribute: The string the attribute whose value is to be returned.
     The string is represented as `:namespace/name`.
     - Parameter entid: The `Entid` of the entity we want the value from.

     - Returns: The `TypedValue` containing the value of the attribute for the entity.
     */
    open func value(forAttribute attribute: String, ofEntity entid: Entid) throws -> TypedValue? {
        return TypedValue(raw: try RustError.unwrap({err in
            store_value_for_attribute(self.raw, entid, attribute, err)
        }));
    }

    // Destroys the pointer by passing it back into Rust to be cleaned up
    override open func cleanup(pointer: OpaquePointer) {
        store_destroy(pointer)
    }
}

/**
 Set up `Mentat` to provide an interface to Mentat's transaction observation
 */
extension Mentat: Observable {
     /**
     Register an `Observing` and a set of attributes to observer for transaction observation.
     The `transactionDidOccur(String: [TxChange]:)` function is called when a transaction
     occurs in the `Store` that this `Mentat` is connected to that affects the attributes that an
     `Observing` has registered for.

     - Parameter key: `String` representing an identifier for the `Observing`.
     - Parameter observer: The `Observing` to be notified when a transaction occurs.
     - Parameter attributes: An `Array` of `Strings` representing the attributes that the `Observing`
     wishes to be notified about if they are referenced in a transaction.
     */
    open func register(key: String, observer: Observing, attributes: [String]) {
        let attrEntIds = attributes.map({ (kw) -> Entid in
            let entid = Entid(self.entidForAttribute(attribute: kw));
            return entid
        })

        let ptr = UnsafeMutablePointer<Entid>.allocate(capacity: attrEntIds.count)
        let entidPointer = UnsafeMutableBufferPointer(start: ptr, count: attrEntIds.count)
        var _ = entidPointer.initialize(from: attrEntIds)

        guard let firstElement = entidPointer.baseAddress else {
            return
        }
        Mentat.observers[key] = observer
        store_register_observer(self.raw, key, firstElement, Entid(attributes.count), transactionObserverCallback)

    }

    /**
     Unregister the `Observing` that was registered with the provided key such that it will no longer be called
     if a transaction occurs that affects the attributes that `Observing` was registered to observe.

     The `Observing` will need to re-register if it wants to start observing again.

     - Parameter key: `String` representing an identifier for the `Observing`.
     */
    open func unregister(key: String) {
        Mentat.observers.removeValue(forKey: key)
        store_unregister_observer(self.raw, key)
    }
}


/**
 This function needs to be static as callbacks passed into Rust from Swift cannot contain state. Therefore the observers are static, as is
 the function that we pass into Rust to receive the callback.
 */
private func transactionObserverCallback(key: UnsafePointer<CChar>, reports: UnsafePointer<TxChangeList>) {
    let key = String(cString: key)
    guard let observer = Mentat.observers[key] else { return }
    DispatchQueue.global(qos: .background).async {
        observer.transactionDidOccur(key: key, reports: [TxChange]())
    }
}

File: sdks/android/Mentat/library/src/main/java/com/mozilla/mentat/Mentat.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * Copyright 2018 Mozilla
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License. */

package com.mozilla.mentat;

import android.util.Log;

import com.sun.jna.Memory;
import com.sun.jna.Pointer;

/**
 * The primary class for accessing Mentat's API.<br/>
 * This class provides all of the basic API that can be found in Mentat's Store struct.<br/>
 * The raw pointer it holds is a pointer to a Store.
 */
public class Mentat extends RustObject {

    static {
        System.loadLibrary("mentat_ffi");
    }

    /**
     * Open a connection to a Store in a given location.<br/>
     * If the store does not already exist, one will be created.
     * @param dbPath    The URI as a String of the store to open.
     */
    public Mentat(String dbPath) {
        this.rawPointer = JNA.INSTANCE.store_open(dbPath);
    }

    /**
     * Open a connection to an in-memory Store.
     */
    public Mentat() {
        this.rawPointer = JNA.INSTANCE.store_open("");
    }

    /**
     * Create a new Mentat with the provided pointer to a Mentat Store
     * @param rawPointer    A pointer to a Mentat Store.
     */
    public Mentat(Pointer rawPointer) { this.rawPointer = rawPointer; }

    /**
     * Add an attribute to the cache. The {@link CacheDirection} determines how that attribute can be
     * looked up.
     *
     * TODO: Throw an exception if cache action fails
     *
     * @param attribute The attribute to cache
     * @param direction The direction the attribute should be keyed.
     * FORWARD caches values for an attribute keyed by entity
     * (i.e. find values and entities that have this attribute, or find values of attribute for an entity)
     * REVERSE caches entities for an attribute keyed by value.
     * (i.e. find entities that have a particular value for an attribute).
     * BOTH adds an attribute such that it is cached in both directions.
     */
    public void cache(String attribute, CacheDirection direction) {
        RustResult result = null;
        switch (direction) {
            case FORWARD:
                result = JNA.INSTANCE.store_cache_attribute_forward(this.rawPointer, attribute);
            case REVERSE:
                result = JNA.INSTANCE.store_cache_attribute_reverse(this.rawPointer, attribute);
            case BOTH:
                result = JNA.INSTANCE.store_cache_attribute_bi_directional(this.rawPointer, attribute);
        }
        if (result.isFailure()) {
            Log.e("Mentat", result.err);
        }
    }

    /**
     * Simple transact of an EDN string.
     * TODO: Throw an exception if the transact fails
     * @param transaction   The string, as EDN, to be transacted.
     * @return  The {@link TxReport} of the completed transaction
     */
    public TxReport transact(String transaction) {
        RustResult result = JNA.INSTANCE.store_transact(this.rawPointer, transaction);
        if (result.isFailure()) {
            Log.e("Mentat", result.err);
            return null;
        }

        if (result.isSuccess()) {
            return new TxReport(result.ok);
        } else {
            return null;
        }
    }

    /**
     * Get the the `Entid` of the attribute
     * @param attribute The string represeting the attribute whose `Entid` we are after. The string is represented as `:namespace/name`.
     * @return  The `Entid` associated with the attribute.
     */
    public long entIdForAttribute(String attribute) {
        return JNA.INSTANCE.store_entid_for_attribute(this.rawPointer, attribute);
    }

    /**
     * Start a query.
     * @param query The string represeting the the query to be executed.
     * @return  The {@link Query} representing the query that can be executed.
     */
    public Query query(String query) {
        return new Query(JNA.INSTANCE.store_query(this.rawPointer, query));
    }

    /**
     * Retrieve a single value of an attribute for an Entity
     * TODO: Throw an exception if the result contains an error.
     * @param attribute The string the attribute whose value is to be returned. The string is represented as `:namespace/name`.
     * @param entid The `Entid` of the entity we want the value from.
     * @return  The {@link TypedValue} containing the value of the attribute for the entity.
     */
    public TypedValue valueForAttributeOfEntity(String attribute, long entid) {
        RustResult result = JNA.INSTANCE.store_value_for_attribute(this.rawPointer, entid, attribute);

        if (result.isSuccess()) {
            return new TypedValue(result.ok);
        }

        if (result.isFailure()) {
            Log.e("Mentat", result.err);
        }

        return null;
    }

    /**
     * Register an callback and a set of attributes to observer for transaction observation.
     * The callback function is called when a transaction occurs in the `Store` that this `Mentat`
     * is connected to that affects the attributes that an observer has registered for.
     * @param key   `String` representing an identifier for the observer.
     * @param attributes    An array of Strings representing the attributes that the observer wishes
     *                     to be notified about if they are referenced in a transaction.
     * @param callback  the function to call when an observer notice is fired.
     */
    public void registerObserver(String key, String[] attributes, TxObserverCallback callback) {
        // turn string array into int array
        long[] attrEntids = new long[attributes.length];
        for(int i = 0; i < attributes.length; i++) {
            attrEntids[i] = JNA.INSTANCE.store_entid_for_attribute(this.rawPointer, attributes[i]);
        }
        final Pointer entidsNativeArray = new Memory(8 * attrEntids.length);
        entidsNativeArray.write(0, attrEntids, 0, attrEntids.length);
        JNA.INSTANCE.store_register_observer(rawPointer, key, entidsNativeArray, attrEntids.length, callback);
    }

    /**
     * Unregister the observer that was registered with the provided key such that it will no longer be called
     * if a transaction occurs that affects the attributes that the observer was registered to observe.
     * <p/>
     * The observer will need to re-register if it wants to start observing again.
     * @param key   String representing an identifier for the observer.
     */
    public void unregisterObserver(String key) {
        JNA.INSTANCE.store_unregister_observer(rawPointer, key);
    }


    /**
     * Start a new transaction
     *
     * TODO: Throw an exception if the result contains an error.
     *
     * @return The {@link InProgress} used to manage the transaction
     */
    public InProgress beginTransaction() {
        RustResult result = JNA.INSTANCE.store_begin_transaction(this.rawPointer);
        if (result.isSuccess()) {
            return new InProgress(result.ok);
        }

        if (result.isFailure()) {
            Log.i("Mentat", result.err);
        }

        return null;
    }

    /**
     * Creates a new transaction ({@link InProgress}) and returns an {@link InProgressBuilder} for
     * that transaction.
     *
     * TODO: Throw an exception if the result contains an error.
     *
     * @return  an {@link InProgressBuilder} for a new transaction.
     */
    public InProgressBuilder entityBuilder() {
        RustResult result = JNA.INSTANCE.store_in_progress_builder(this.rawPointer);
        if (result.isSuccess()) {
            return new InProgressBuilder(result.ok);
        }

        if (result.isFailure()) {
            Log.i("Mentat", result.err);
        }

        return null;
    }

    /**
     * Creates a new transaction ({@link InProgress}) and returns an {@link EntityBuilder} for the
     * entity with `entid` for that transaction.
     *
     * TODO: Throw an exception if the result contains an error.
     *
     * @param entid The `Entid` for this entity.
     * @return  an {@link EntityBuilder} for a new transaction.
     */
    public EntityBuilder entityBuilder(long entid) {
        RustResult result = JNA.INSTANCE.store_entity_builder_from_entid(this.rawPointer, entid);
        if (result.isSuccess()) {
            return new EntityBuilder(result.ok);
        }

        if (result.isFailure()) {
            Log.i("Mentat", result.err);
        }

        return null;
    }

    /**
     * Creates a new transaction ({@link InProgress}) and returns an {@link EntityBuilder} for a new
     * entity with `tempId` for that transaction.
     *
     * TODO: Throw an exception if the result contains an error.
     *
     * @param tempId    The temporary identifier for this entity.
     * @return  an {@link EntityBuilder} for a new transaction.
     */
    public EntityBuilder entityBuilder(String tempId) {
        RustResult result = JNA.INSTANCE.store_entity_builder_from_temp_id(this.rawPointer, tempId);
        if (result.isSuccess()) {
            return new EntityBuilder(result.ok);
        }

        if (result.isFailure()) {
            Log.i("Mentat", result.err);
        }

        return null;
    }

    @Override
    public void close() {
        if (this.rawPointer != null) {
            JNA.INSTANCE.store_destroy(this.rawPointer);
        }
    }
}
